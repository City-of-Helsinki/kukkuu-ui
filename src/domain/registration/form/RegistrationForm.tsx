import React, { FunctionComponent } from 'react';
import { Formik, Field } from 'formik';
import { connect } from 'react-redux';
import { useMutation } from '@apollo/react-hooks';
import { useTranslation } from 'react-i18next';

import styles from './registrationForm.module.scss';
import Button from '../../../common/components/button/Button';
import InputField from '../../../common/components/form/fields/input/InputField';
import SelectField from '../../../common/components/form/fields/select/SelectField';
import submitChildMutationQuery from '../mutations/submitChild';
import { setFormValues } from '../state/RegistrationActions';
import { RegistrationFormValues } from '../types/RegistrationTypes';
import { StoreState } from '../../app/types/AppTypes';
import { registrationFormDataSelector } from '../state/RegistrationSelectors';
import { convertFormValuesToSubmitChildMutation } from './RegistrationFormUtils';
import { formatTime, newMoment } from '../../../common/time/utils';
import { DEFAULT_DATE_FORMAT } from '../../../common/time/TimeConstants';

interface Props {
  setFormValues: (values: RegistrationFormValues) => void;
  initialValues: RegistrationFormValues;
}

const RegistrationForm: FunctionComponent<Props> = ({
  setFormValues,
  initialValues,
}) => {
  const handleSubmit = (values: RegistrationFormValues) => {
    // eslint-disable-next-line @typescript-eslint/camelcase
    //const submitValues: SubmitChildValues = convertFormValuesToSubmitChildMutation(
    //  values
    //);
  };
  const [submitChild, { data }] = useMutation(submitChildMutationQuery);
  console.log(data);
  const { t } = useTranslation();

  return (
    <div className={styles.registrationForm}>
      <Formik
        initialValues={initialValues}
        onSubmit={e => {
          setFormValues(e);
          const z = convertFormValuesToSubmitChildMutation(e);
          console.log(z);
          try {
            submitChild({
              variables: z,
            });
          } catch (err) {
            console.log(err);
          }
        }}
      >
        {({ values, handleChange, isSubmitting, handleSubmit, isValid }) => (
          <form onSubmit={handleSubmit}>
            <div className={styles.childInfo}>
              <Field
                type="text"
                name="child.birthday"
                label={t('registration.form.child.birthday.input.label')}
                onChange={handleChange}
                value={formatTime(
                  newMoment(values.child.birthday),
                  DEFAULT_DATE_FORMAT
                )}
                component={InputField}
                disabled={true}
                placeholder={t(
                  'registration.form.child.birthday.input.placeholder'
                )}
              />
              <Field
                type="text"
                name="child.homeCity"
                label={t('registration.form.child.homeCity.input.label')}
                onChange={handleChange}
                value={values.child.homeCity}
                component={InputField}
                placeholder={t(
                  'registration.form.child.homeCity.input.placeholder'
                )}
              />
              <div className={styles.childName}>
                <Field
                  type="text"
                  name="child.firstName"
                  label={t('registration.form.child.firstName.input.label')}
                  onChange={handleChange}
                  value={values.child.firstName}
                  component={InputField}
                  placeholder={t(
                    'registration.form.child.firstName.input.placeholder'
                  )}
                />
                <Field
                  type="text"
                  name="child.lastName"
                  label={t('registration.form.child.lastName.input.label')}
                  onChange={handleChange}
                  value={values.child.lastName}
                  component={InputField}
                  placeholder={t(
                    'registration.form.child.lastName.input.placeholder'
                  )}
                />
              </div>
            </div>

            <div className={styles.guardianInfo}>
              <Field
                type="text"
                name="guardian.email"
                label={t('registration.form.guardian.email.input.label')}
                onChange={handleChange}
                value={values.guardian.email}
                component={InputField}
                placeholder={t(
                  'registration.form.guardian.email.input.placeholder'
                )}
              />

              <Field
                type="text"
                name="guardian.phoneNumber"
                label={t('registration.form.guardian.phoneNumber.input.label')}
                onChange={handleChange}
                value={values.guardian.phoneNumber}
                component={InputField}
                placeholder={t(
                  'registration.form.guardian.phoneNumber.input.placeholder'
                )}
              />
              <div className={styles.guardianName}>
                <Field
                  type="text"
                  name="guardian.firstName"
                  label={t('registration.form.guardian.firstName.input.label')}
                  onChange={handleChange}
                  value={values.guardian.firstName}
                  component={InputField}
                  placeholder={t(
                    'registration.form.guardian.firstName.input.placeholder'
                  )}
                />
                <Field
                  type="text"
                  name="guardian.lastName"
                  label={t('registration.form.guardian.lastName.input.label')}
                  onChange={handleChange}
                  value={values.guardian.lastName}
                  component={InputField}
                  placeholder={t(
                    'registration.form.guardian.lastName.input.placeholder'
                  )}
                />
              </div>
              <Field
                name="guardian.relationship"
                label={t('registration.form.guardian.relationship.input.label')}
                onChange={handleChange}
                value={values.guardian.relationship}
                component={SelectField}
                id="registration.form.guardian.relationship.select"
                options={[
                  { label: 'Parents', value: 'parents' },
                  { label: 'Spouse', value: 'spouse' },
                ]}
                placeholder={t(
                  'registration.form.guardian.relationship.input.placeholder'
                )}
              />
              <Field
                type="checkbox"
                checked={values.verifyInformation}
                name="verifyInformation"
                label={t('registration.form.verifyInformation.input.label')}
                onChange={handleChange}
                component={InputField}
                value={values.verifyInformation}
              />
            </div>

            <Button
              type="submit"
              className={styles.submitButton}
              disabled={isSubmitting || !isValid}
            >
              {t('homePage.hero.buttonText')}
            </Button>
          </form>
        )}
      </Formik>
    </div>
  );
};

const actions = {
  setFormValues,
};

const mapStateToProps = (state: StoreState) => ({
  initialValues: registrationFormDataSelector(state),
});

export const UnconnectedRegistrationForm = RegistrationForm;

export default connect(
  mapStateToProps,
  actions
)(RegistrationForm);
