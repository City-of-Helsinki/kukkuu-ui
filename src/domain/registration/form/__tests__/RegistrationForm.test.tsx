import RegistrationForm, {
  EMAIL_FIELD_TESTID,
  FORM_TESTID,
} from '../RegistrationForm';
import { render, screen } from '../../../../common/test/testingLibraryUtils';

// TODO: Needs mocks for profile query and redux selector.
it.skip('renders snapshot correctly', async () => {
  const { container } = render(<RegistrationForm />);
  expect(await screen.findByTestId(FORM_TESTID));
  expect(container).toMatchSnapshot();
});

it('email field is disabled', async () => {
  render(<RegistrationForm />);
  const emailField = await screen.findByTestId(EMAIL_FIELD_TESTID);
  expect(emailField.hasAttribute('disabled'));
});
