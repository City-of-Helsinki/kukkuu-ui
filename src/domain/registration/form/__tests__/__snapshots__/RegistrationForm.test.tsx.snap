// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `"<div class=\\"registrationForm\\"><form><div class=\\"childInfo\\"><div class=\\"childName\\"><div class=\\"inputField\\"><div class=\\"inputWrapper textInput\\"><div class=\\"inputTypeWrapper\\"><label>Child first name</label><input type=\\"text\\" name=\\"child.firstName\\" value=\\"\\" placeholder=\\"Please put child first name\\"/></div></div><div class=\\"inputErrorMessage\\"></div></div><div class=\\"inputField\\"><div class=\\"inputWrapper textInput\\"><div class=\\"inputTypeWrapper\\"><label>Child last name</label><input type=\\"text\\" name=\\"child.lastName\\" value=\\"\\" placeholder=\\"Please put child last name\\"/></div></div><div class=\\"inputErrorMessage\\"></div></div></div></div><div class=\\"guardianInfo\\"><div class=\\"inputField\\"><div class=\\"inputWrapper textInput\\"><div class=\\"inputTypeWrapper\\"><label>Guardian email</label><input type=\\"text\\" name=\\"guardian.email\\" value=\\"\\" placeholder=\\"Please put guardian email\\"/></div></div><div class=\\"inputErrorMessage\\"></div></div><div class=\\"inputField\\"><div class=\\"inputWrapper textInput\\"><div class=\\"inputTypeWrapper\\"><label>Guardian phone number</label><input type=\\"text\\" name=\\"guardian.phoneNumber\\" value=\\"\\" placeholder=\\"Please put guardian phone number\\"/></div></div><div class=\\"inputErrorMessage\\"></div></div><div class=\\"guardianName\\"><div class=\\"inputField\\"><div class=\\"inputWrapper textInput\\"><div class=\\"inputTypeWrapper\\"><label>Guardian first name</label><input type=\\"text\\" name=\\"guardian.firstName\\" value=\\"\\" placeholder=\\"Please put guardian first name\\"/></div></div><div class=\\"inputErrorMessage\\"></div></div><div class=\\"inputField\\"><div class=\\"inputWrapper textInput\\"><div class=\\"inputTypeWrapper\\"><label>Guardian last name</label><input type=\\"text\\" name=\\"guardian.lastName\\" value=\\"\\" placeholder=\\"Please put guardian last name\\"/></div></div><div class=\\"inputErrorMessage\\"></div></div></div><div class=\\"selectWrapper\\"><label for=\\"registration.form.guardian.relationship.select\\">Guardian relation with child</label><select id=\\"registration.form.guardian.relationship.select\\" field=\\"[object Object]\\" placeholder=\\"Please put guardian relation\\"><option selected=\\"\\" value=\\"parents\\">Parents</option><option value=\\"spouse\\">Spouse</option></select></div><div class=\\"inputField\\"><div class=\\"inputWrapper checkboxInput\\"><div class=\\"inputTypeWrapper\\"><input type=\\"checkbox\\" id=\\"agree\\" value=\\"false\\"/><label for=\\"agree\\">registration.form.agree.input.label</label></div></div><div class=\\"inputErrorMessage\\"></div></div></div><div class=\\"submitButton buttonWrapper\\"><button type=\\"submit\\">Ilmoittaudu mukaan</button></div></form></div>"`;
